Networking/Networking/ProductImageInBackground/ProductImageStatusStorage.swift

```diff
@@ -0,0 +1,183 @@
+                    return (siteID: status.siteID,
+                            productOrVariationID: status.productOrVariationID,
+                            assetType: status.assetType,
+                            error: error)
+                }
+                return nil
+            }
+            .eraseToAnyPublisher()
+    }
+
+    public init(userDefaults: UserDefaults = .standard,
+         key: String = "savedProductUploadImageStatuses",
+         encoder: JSONEncoder = JSONEncoder(),
+         decoder: JSONDecoder = JSONDecoder()) {
+        self.key = key
+        self.userDefaults = userDefaults
+        self.encoder = encoder
+        self.decoder = decoder
+        self.encoder.dateEncodingStrategy = .iso8601
+        self.decoder.dateDecodingStrategy = .iso8601
+        self.statusesSubject = CurrentValueSubject(Self.loadStatuses(from: userDefaults, key: key, decoder: decoder))
+        setupObservers()
+    }
+
+    public func addStatus(_ status: ProductImageStatus) {
+        var current = statusesSubject.value
+        current.append(status)
+        saveStatuses(current)
+    }
+
+    public func removeStatus(_ status: ProductImageStatus) {
+        let current = statusesSubject.value.filter { $0 != status }
+        saveStatuses(current)
+    }
+
+    public func removeStatus(where predicate: (ProductImageStatus) -> Bool) {
+        let current = statusesSubject.value.filter { !predicate($0) }
+        saveStatuses(current)
+    }
+
+    public func updateStatus(_ status: ProductImageStatus) {
+        var current = statusesSubject.value
+        if let index = current.firstIndex(where: { $0 == status }) {
+            current[index] = status
+            saveStatuses(current)
+        } else {
+            addStatus(status)
+        }
+    }
+
+    public func findStatus(where predicate: (ProductImageStatus) -> Bool) -> ProductImageStatus? {
+        statusesSubject.value.first(where: predicate)
+    }
+
+    public func getAllStatuses() -> [ProductImageStatus] {
+        statusesSubject.value
+    }
+
+    public func getAllStatuses(for siteID: Int64, productID: ProductOrVariationID?) -> [ProductImageStatus] {
+        statusesSubject.value.filter {
+            switch $0 {
+            case .uploading(_, let sID, let pID),
+                    .uploadFailure(_, _, let sID, let pID):
+                return sID == siteID && (productID == nil || pID == productID)
+            case .remote(_, let sID, let pID):
+                return sID == siteID && pID == productID
+            }
+        }
+    }
+
+    public func clearAllStatuses() {
+        userDefaults.removeObject(forKey: key)
+        userDefaults.synchronize()
+        statusesSubject.send([])
+    }
+
+    public func setAllStatuses(_ statuses: [ProductImageStatus]) {
+        saveStatuses(statuses)
+    }
+
+    public func setAllStatuses(_ statuses: [ProductImageStatus], for siteID: Int64, productID: ProductOrVariationID?) {
```

Comment:
Should this be renamed to `appendStatuses` to differentiate with `setAllStatuses` above?

Reply:
It makes sense and is less confusing. ⚡

Reply:
I implemented the changes in this commit: 016d77740fd7e27bc6d0604e41c898ffbdf1073d